feat: Add RateLimiter in AbstractObservable

srv-twry opened the pull request:
Fixes #1037 
Fixes #1041 

Checklist:

- [x] I have checked for PMD and check-style issues <!-- please add a note if a false warning could not be suppressed -->
- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream development branch.

Changes: 
- Add RateLimiter in AbstractObservable.
- Fix infinite loops in Tracks/Sessions/Sponsors list.

Screenshots for the change: N/A

srv-twry commented on app/src/main/java/org/fossasia/openevent/app/data/RateLimiter.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Utility class that decides whether we should fetch some data or not.
+ */
+public class RateLimiter<KEY> {
```
@iamareebjamal 
Should I change it to `K` or leave it ? It has been taken directly from the android samples.

srv-twry commented:
@iamareebjamal Is it the right way according to you ? The build fails since I haven't updated the tests. I'll do it once we agree that this is the right method.

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/RateLimiter.java:
```diff
@@ -0,0 +1,40 @@
+import android.os.SystemClock;
+import android.support.v4.util.ArrayMap;
+
+import java.util.concurrent.TimeUnit;
```
Will crash on lower APIs

srv-twry replied:
Will Duration work ? https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html
We are using Java8 so this shouldn't cause problems. Am I right ?

iamareebjamal replied:
ART != JDK

Android supports Java 8 Time API from API 24+

srv-twry replied:
Using ThreeTen back port for Duration should work. http://www.threeten.org/threetenbp/apidocs/org/threeten/bp/Duration.html

iamareebjamal replied:
Yes, you can use TimeUnit too if it is backported

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/AbstractObservable.java:
```diff
@@ -18,15 +19,19 @@
         private Observable<T> networkObservable;
 
-        public AbstractObservableBuilder(ConnectionStatus connectionStatus) {
+        public AbstractObservableBuilder(ConnectionStatus connectionStatus, AbstractObservable abstractObservable) {
```
Since we are building abstract observable, I don't think it'd be good to ask it in constructor

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/RateLimiter.java:
```diff
@@ -0,0 +1,40 @@
+ * Utility class that decides whether we should fetch some data or not.
+ */
+public class RateLimiter<KEY> {
+    private ArrayMap<KEY, Long> timestamps = new ArrayMap<>();
```
Variable type should be of map

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/AbstractObservable.java:
```diff
@@ -18,15 +19,19 @@
 public final class AbstractObservable {
 
     private final ConnectionStatus connectionStatus;
+    private final RateLimiter<String> rateLimiter = new RateLimiter<>(Duration.ofMinutes(10));
```
Also, about this rate limiter, will it not disallow people to pull to refresh?

srv-twry replied:
No it won't, An API call is made in two cases: 
1. `reload` is true.
2. An API call wasn't made in the last specified duration(10 minutes in my case).

``` Java
if (reload || abstractObservable.rateLimiter.shouldFetch(rateLimiterKey)) {
                    return networkObservable
                        .doOnNext(item -> Timber.d("Loaded %s From Network on Thread %s",
                            item.getClass(), Thread.currentThread().getName()));
                }
```

iamareebjamal replied:
OK, thanks

iamareebjamal replied:
Please change the builder constructor and fix the tests

srv-twry commented on app/src/main/java/org/fossasia/openevent/app/data/AbstractObservable.java:
```diff
@@ -35,6 +39,18 @@ public AbstractObservableBuilder(ConnectionStatus connectionStatus) {
+            this.abstractObservable = abstractObservable;
+
+            return this;
+        }
```
Any suggestions on how I could accomplish it without asking in the constructor ? 
Making the rate limiter static in AbstractObservable ?

``` Java
private static final RateLimiter<String> rateLimiter = new RateLimiter<>(Duration.ofMinutes(10));
```

and then using inside the AbstractObservableBuilder.

srv-twry replied:
I am adding it in the AbstractObservable only.
I assume you meant `AbstractObservableBuilder` like this:
``` Java
 public static final class AbstractObservableBuilder<T> {
        ...
        private Observable<T> networkObservable;
        private final RateLimiter<String> rateLimiter = new RateLimiter<>(Duration.ofMinutes(10));
```

This won't work since every time we request the repository for data, it creates a new instance of `AbstractObservableBuilder`.
Even if I make the request after 1 sec, it will build a new AbstractObservable and hence a  new rate limiter. The previous one will be gone.

iamareebjamal replied:
Sorry, it seems there has been a mistake in reviewing this. I'll re-review soon

srv-twry commented:
This will work fine.
```Java
public final class AbstractObservable {

    private final ConnectionStatus connectionStatus;
    private static final RateLimiter<String> rateLimiter = new RateLimiter<>(Duration.ofMinutes(10));

    public static final class AbstractObservableBuilder<T> {
        private final ConnectionStatus connectionStatus;
        private boolean reload;
        private String rateLimiterKey;
        private Observable<T> diskObservable;
        private Observable<T> networkObservable;

        public AbstractObservableBuilder(ConnectionStatus connectionStatus) {
            this.connectionStatus = connectionStatus;
        }

        public AbstractObservableBuilder<T> reload(boolean reload) {
            this.reload = reload;

            return this;
        }

        public AbstractObservableBuilder<T> rateLimiter(String rateLimiterKey) {
            this.rateLimiterKey = rateLimiterKey;

            return this;
        }

       ........

        @NonNull
        private Observable<T> getConnectionObservable() {
            if (connectionStatus.isConnected()) {
                if (reload || AbstractObservable.rateLimiter.shouldFetch(rateLimiterKey)) {
                    return networkObservable
                        .doOnNext(item -> Timber.d("Loaded %s From Network on Thread %s",
                            item.getClass(), Thread.currentThread().getName()));
                } else {
                    return Observable.empty();
                }
            }
            else {
                AbstractObservable.rateLimiter.reset(rateLimiterKey);
                return Observable.error(new Throwable(Constants.NO_NETWORK));
            }
        }
        ..........
    }

    @Inject
    public AbstractObservable(ConnectionStatus connectionStatus) {
        this.connectionStatus = connectionStatus;
    }

    public <T> AbstractObservableBuilder<T> of(Class<T> clazz) {
        return new AbstractObservableBuilder<>(connectionStatus);
    }
}

```

srv-twry commented on app/src/main/java/org/fossasia/openevent/app/data/AbstractObservable.java:
```diff
@@ -18,15 +19,19 @@
         private Observable<T> networkObservable;
 
-        public AbstractObservableBuilder(ConnectionStatus connectionStatus) {
+        public AbstractObservableBuilder(ConnectionStatus connectionStatus, RateLimiter<String> rateLimiter) {
```
@iamareebjamal I am taking the RateLimiter in the constructor and storing it inside the AbstractObservableBuilder. Is it fine ?

srv-twry replied:
Will update tests accordingly if this is fine.

iamareebjamal replied:
This constructor takes connection status just because, there is no reason for this, so, if the rate limiter is taken through the builder methods, this can be avoided

Builders should have no or at most 1 constructor parameter

srv-twry commented:
@iamareebjamal Please review. I have updated the PR according to the android architecture samples. They maintain a rate limiter per repository. So I have also added separate rate limiters for every repository.

Example: https://github.com/googlesamples/android-architecture-components/blob/b1a194c1ae267258cd002e2e1c102df7180be473/GithubBrowserSample/app/src/main/java/com/android/example/github/repository/RepoRepository.java

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/attendee/AttendeeRepositoryImpl.java:
```diff
@@ -43,6 +46,8 @@ public AttendeeRepositoryImpl(Repository repository, AttendeeApi attendeeApi) {
         return repository.observableOf(Attendee.class)
             .reload(reload)
+            .rateLimiter(rateLimiter)
+            .rateLimiterKey("Attendee")
```
These both arguments are coupled, take them using one function in builder in order key, rateLimiter

Nice job, LGTM otherwise

srv-twry commented:
@iamareebjamal Updated PR, please review.

iamareebjamal commented on app/src/main/java/org/fossasia/openevent/app/data/AbstractObservable.java:
```diff
@@ -35,6 +47,13 @@ public AbstractObservableBuilder(ConnectionStatus connectionStatus) {
             return this;
         }
 
+        public AbstractObservableBuilder<T> withRateLimiterConfig(RateLimiterConfig rateLimiterConfig) {
```
No need to create a new object, take 2 parameters instead
